v1/tarball/write.go

```diff
@@ -29,5 +32,89 @@ type WriteOptions struct {
 
 // Write saves the image as the given tag in a tarball at the given path.
 func Write(p string, tag name.Tag, img v1.Image, wo WriteOptions) error {
-	return fmt.Errorf("NYI: tarball.Write(%v)", p)
+	// Write in the compressed format.
+	// This is a tarball, on-disk, with:
+	// One manifest.json file at the top level containing information about several images.
+	// One file for each layer, named after the layer's SHA.
+	// One file for the config blob, named after it's sha.
+
+	w, err := os.OpenFile(p, os.O_WRONLY, 0644)
+	if err != nil {
+		return err
+	}
+	defer w.Close()
+
+	tf := tar.NewWriter(w)
+	defer tf.Close()
+
+	// Write the config.
+	cfgName, err := img.ConfigName()
+	if err != nil {
+		return err
+	}
+	cfg, err := img.ConfigFile()
+	if err != nil {
+		return err
+	}
+	cfgBlob, err := json.Marshal(cfg)
+	if err != nil {
+		return err
+	}
+	if err := writeFile(tf, cfgName.String(), cfgBlob); err != nil {
+		return err
+	}
+
+	// Generate the tar descriptor and write it.
+	layerPaths := []string{}
+	layers, err := img.FSLayers()
+	if err != nil {
+		return err
+	}
+	for _, l := range layers {
+		layerPaths = append(layerPaths, l.String())
+	}
+	td := tarDescriptor{
+		singleImageTarDescriptor{
+			Config:   cfgName.String(),
+			RepoTags: []string{tag.String()},
+			Layers:   layerPaths,
+		},
+	}
+	tdBytes, err := json.Marshal(td)
+	if err != nil {
+		return err
+	}
+	if err := writeFile(tf, "manifest.json", tdBytes); err != nil {
+		return err
+	}
+
+	// Write the layers.
+	for _, layer := range layers {
+		r, err := img.Blob(layer)
+		if err != nil {
+			return err
+		}
+		blob, err := ioutil.ReadAll(r)
+		if err != nil {
+			return err
+		}
+		if err := writeFile(tf, layer.String(), blob); err != nil {
```

Comment:
`writeFile` should take an `io.Reader` to avoid having to buffer the entire layer contents. For the JSON blobs it's not so bad, those can be wrapped in a `bytes.NewReader`, but layers can be pretty big.

Reply:
I started with that, but we need to write the Header first which needs to know the size, so it has to get buffered once.

But I think that we can get the BlobSize first and also pass that in. It has to get buffered once during manifest creation, but you're right that we shouldn't have to do it a second time.